rfc/rfc-82/rfc-82.md

```diff
@@ -0,0 +1,254 @@
+```
+Notes:
+- Txn1 operates entirely under schema S1.
+- No schema changes occur; no conflicts arise.
+
+**Scenario 5**
+```
+Time ---------------------------------------------------------------------------------->
+
+Txn1:    [ Start Txn1 ]------------------------------[ Validate ]--------------------[ Commit Txn1 ]
+            Uses Schema: S2                                                     Writes Schema: S2
+
+Table Schema:
+          [ S1 ]------------------------------------[ S1 ]----------------------------[ S2 ]
+```
+Notes:
+- Txn1 starts with schema S1 and evolves it to S2 within the transaction.
+- No other transactions interfere; no conflicts occur.
+
+**Scenario 6**
+```
+Time ---------------------------------------------------------------------------------->
+
+Txn2:           [ Start Txn2 ]------------------------[ Commit Txn2 ]
+                   Evolves Schema: S1 to S2
+
+Txn1:    [ Start Txn1 ]------------------------------[ Validate ]--------------------[ Commit Txn1 ]
+            Uses Schema: S1                                                     Writes Schema: S2
+
+Table Schema:
+          [ S1 ]---------------------------------------[ S2 ]---------------------------------[ S2 ]
+```
+Notes:
+- Txn1 starts with schema S1.
+- Txn2 commits before Txn1 validates, evolving the schema to S2.
+- Txn1 validates against schema S2; backward compatibility allows it to proceed.
+- Txn1 writes data compatible with S2; commits successfully.
+
+**Scenario 7**
+```
+Time ---------------------------------------------------------------------------------->
+
+Txn2:           [ Start Txn2 ]------------------------[ Commit Txn2 ]
+                   Evolves Schema: S1 to S2
+
+Txn1:    [ Start Txn1 ]------------------------------[ Validate ]--------------------[ Commit Txn1 ]
+            Uses Schema: S2                                                     Writes Schema: S2
+
+Table Schema:
+          [ S1 ]--------------------------[ S2 ]----------------------------------------[ S2 ]
+```
+Notes:
+- Both Txn1 and Txn2 aim to evolve schema from S1 to S2.
+- Txn2 commits before Txn1 validates, updating schema to S2.
+- Txn1 detects that schema is already S2; no conflict occurs.
+- Txn1 commits successfully.
+
+**Scenario 8**
+```
+Time ---------------------------------------------------------------------------------->
+
+Txn2:           [ Start Txn2 ]------------------------[ Commit Txn2 ]
+                   Evolves Schema: S1 to S2
+
+Txn1:    [ Start Txn1 ]------------------------------[ Validate ]----X
+            Uses Schema: S3
+
+Table Schema:
+          [ S1 ]--------------------------------------[ S2 ]-------------------------------------[S2]
+```
+Notes:
+- Txn1 intends to evolve schema from S1 to S3.
+- Txn2 commits before Txn1 validates, updating schema to S2.
+- At validation, Txn1 detects schema conflict (S2 vs. S3); transaction fails.
```

Comment:
In summry, it looks like there is a simple rule for all these scenarios: validate the latest schema again the write schema during conflict resolution. There is no need to validate the start schema (before a txn) and the write schema because writer would reconcile that.

But currently we do writer schema evolving in each executors, how about different executors use different schema as the base schema to evolve then?

Reply:
for your first comment - we need to validate start schema to differentiate case 2 and case 6. case 6 implicitly relying on the existing schema evolution check.

> how about different executors use different schema as the base schema to evolve then?

I might miss some context here. 
- How "base schema" is defined?
- Why executors matter here? We only care 3 schemas here in OCC scenario as mentioned above. How executors breaks the format of the solution?

Reply:
- the base schema is fetched right from the table in executors
- it may not cause error but it may introduce schema in-consistency because different executors uses multiple writer schema to commit, we should keep the compatibility for all these schemas. But generally I kind of think we should unify the write schema at lest in the commit level, all the executors should use one base schema to evolve to the latest write schema.

Reply:
> for your first comment - we need to validate start schema to differentiate case 2 and case 6. case 6 implicitly relying on the existing schema evolution check.

Both case 2 and 6 just validate the write schema against the latest table schema, no matter whether the table schema had been evolved during the write process. 

Still we should calrify this question: how do we manage the granularity of base schemas during a write, it looks like we have a probability of unified write schema but varigated base schemas during one commit write.

Reply:
Discussed with Ethan and Added Q&A section in the end cover the concerns. At least for spark engine, executors uses the same writer schema from the writer config. We don't need to worry about different base schema used by executors

Even if it does we are fine, since the formula of resolving table schema at Txn start time, txn validation time and txn schema does not change. Executors (in spark) even they need to tackle different base schema it is existing behavior and we basically does not make anything worse but get better protection at least in some cases.

Reply:
@danny0405 maybe I miss something on how Flink writer works in this case.  At least the write handle takes the write schema from the write config (set by Spark Datasource and SQL writers, Hudi Streamer on Spark), and the write schema is derived from the input data and table schema at the beginning of the transaction, so the write schema is the same across all executors.  Does Flink writer work in a different way?

Reply:
Either case, as we discussed and @Davis-Zhang-Onehouse mentioned, the proposed concurrent schema evolution detection should be orthogonal.  We can fix any discrepancy on schema across executors if there is any in a follow-up.

Reply:
@Davis-Zhang-Onehouse could you create a follow-up Jira ticket if needed?

Reply:
https://issues.apache.org/jira/browse/HUDI-8332